### PR TITLE
 Strip the PDF.js prefix from URLs (fixes brave/brave-browser#368)

### PR DESCRIPTION
**Pull request by fmarier**
Fixes brave/brave-browser#368.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [ ] macOS
  - [x] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [ ] Windows
  - [ ] macOS
  - [x] Linux
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:

Open <https://fmarier.org/test.pdf> and ensure the URL bar displays `https://fmarier.org/test.pdf`.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

**Comment by fmarier** [comment]
@bridiver, this pull request was re-written following your review comments in https://github.com/brave/brave-core/pull/1074#discussion_r246190933.

Replacing the stripped-down PDF.js extension with the real one was suggested by @bbondy in https://bravesoftware.slack.com/threads/convo/C7VLGSR55-1549589742.278400/. I did some `git blame` digging but couldn't find any explanation for the custom version of PDF.js. The size difference is 2 MB v. 765 KB.

Removing the `pdfjs.pem` file didn't seem to have any effect on any of the existing tests. I'm not entirely sure what its purpose was, though there is a general explanation in https://github.com/brave/brave-core/commit/27287112d58e60934e2fefced98f7c50bb50abe0.

File: browser/extensions/brave_extension_provider_browsertest.cc [hunk 1]
```diff
@@ -65,6 +66,15 @@ IN_PROC_BROWSER_TEST_F(BraveExtensionProviderTest, PDFJSInstalls) {
       "window.domAutomationController.send("
         "!!document.body.querySelector(\"#chrome-pdfjs-logo-bg\"))", &pdfjs_exists));
   ASSERT_TRUE(pdfjs_exists);
+
+  // Ensure that the extension prefix is not in the display URL (brave-browser#368)
+  EXPECT_STREQ(contents->GetController().GetLastCommittedEntry()
+               ->GetVirtualURL().spec().c_str(),
+               url.spec().c_str()); // no "chrome-extension://" prefix
+  std::string internal_url =
+    "chrome-extension://oemmndcbldboiebfnladdacbdfmadadm/" + url.spec();
```
Comment (bridiver) [inline browser/extensions/brave_extension_provider_browsertest.cc]: nit: continuation of a line should have 4 spaces (or more if some kind of alignment with next/previous line helps readability)

File: browser/extensions/brave_extension_provider_browsertest.cc [hunk 2]
```diff
@@ -65,6 +66,15 @@ IN_PROC_BROWSER_TEST_F(BraveExtensionProviderTest, PDFJSInstalls) {
       "window.domAutomationController.send("
         "!!document.body.querySelector(\"#chrome-pdfjs-logo-bg\"))", &pdfjs_exists));
   ASSERT_TRUE(pdfjs_exists);
+
+  // Ensure that the extension prefix is not in the display URL (brave-browser#368)
+  EXPECT_STREQ(contents->GetController().GetLastCommittedEntry()
+               ->GetVirtualURL().spec().c_str(),
```
Comment (bridiver) [inline browser/extensions/brave_extension_provider_browsertest.cc]: also nit: I would actually indent this 4 more spaces to show continuation of the current param vs next param below which can keep the current indentation

File: browser/brave_content_browser_client.cc [hunk 3]
```diff
@@ -115,6 +117,40 @@ bool HandleURLRewrite(GURL* url,
...
+  if (url->SchemeIs(extensions::kExtensionScheme) &&
+      url->DomainIs(pdfjs_extension_id) && url->has_query()) {
+    // The path looks like this:
+    //   /content/web/viewer.html?file=https%3A%2F%2Fexample.com%2Ftest.pdf
+    base::StringPairs params;
+    if (base::SplitStringIntoKeyValuePairs(url->query(), '=', '&', &params)) {
+      for (size_t i = 0; i < params.size(); ++i) {
+        if (params[i].first == "file" && !params[i].second.empty()) {
+          net::UnescapeRule::Type rule =
+            net::UnescapeRule::URL_SPECIAL_CHARS_EXCEPT_PATH_SEPARATORS |
```
Comment (bridiver) [inline browser/brave_content_browser_client.cc]: nit: also continuation of line so should be indented 4 spaces from previous line

File: browser/brave_content_browser_client.cc [hunk 4]
```diff
@@ -115,6 +117,40 @@ bool HandleURLRewrite(GURL* url,
...
+  }
+
+  return false;
+}
+
+bool HandlePDFjsURLReverseRewrite(GURL* url,
+                                  content::BrowserContext* browser_context) {
+  // Remove the PDF.js prefix from display URLs unless that's all there is.
+  if (url->SchemeIs(extensions::kExtensionScheme) &&
+      url->DomainIs(pdfjs_extension_id) && url->has_query()) {
```
Comment (bridiver) [inline browser/brave_content_browser_client.cc]: nit: I would move url->has_query() down to the next line to make it easier to read it as 3 `&&` conditions

File: browser/brave_content_browser_client.cc [hunk 5]
```diff
@@ -17,6 +17,7 @@
 #include "brave/browser/renderer_host/brave_navigation_ui_data.h"
 #include "brave/browser/tor/tor_profile_service_factory.h"
 #include "brave/common/brave_cookie_blocking.h"
+#include "brave/common/extensions/extension_constants.h"
```
Comment (bridiver) [inline browser/brave_content_browser_client.cc]: please put all extension related code in `#if BUILDFLAG(ENABLE_EXTENSIONS)` guards. The headers can go in the existing guard and the new code should be wrapped in `#if BUILDFLAG(ENABLE_EXTENSIONS)` as well

**Comment by fmarier** [comment]
Test URLs:

- HTTP: <http://fmarier.com/test.pdf>
- HTTPS: <https://fmarier.org/test.pdf>

**Comment by bbondy** [comment]
@fmarier I think this should be 0.63.x since the PR is against master. I adjusted it.

**Comment by fmarier** [comment]
This will need to be re-written in light of #2024.

**Comment by bbondy** [comment]
I think this is probably wontfix because of https://github.com/brave/brave-browser/issues/3846 ?

**Comment by fmarier** [comment]
> I think this is probably wontfix because of brave/brave-browser#3846 ?

Indeed. Let's close this.

**Pull request closed without merging**